### PR TITLE
feat: selector picker for when auto-detection is unique over meaningful 

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

The toolbar lets you pick selectors for elements to create actions. The library we use (and others out there) concentrates on choosing a unique selector. But sometimes the unique selector is not the meaningful one.

## unique selector

Screenshot 2023-02-12 at 22 33 03

## meaningful selector

Screenshot 2023-02-12 at 22 42 39

Writing our own selector-detecting library might be the way to go. But there are a lot of abandoned libraries out there so this is probably a hard thing to do that people give up on.

But humans are pretty good at identifying meaningful text...

## Changes

* adds a modal that shows the elements wrapping the selected element
* lets clicking on things build up a selector
* overrides the selector for that element and uses it if creating an action

manual-selectors

## How did you test this code?

* clicking around in the UI and seeing what happens

### CODE REVIEW

File: frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.scss [hunk 1]
```diff
@@ -0,0 +1,15 @@
+.HtmlElements {
```
Comment (mariusandra) [inline frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.scss:1]: Nit:

```suggestion
.HtmlElementsDisplay {
```

File: frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                } catch (e) {
+                    console.error(e)
+                    setSelectorMatches([])
+                }
+            }
+
+            setChosenSelector(builtSelector)
+            onChange?.(builtSelector, checkUniqueness ? selectorMatchCount === 1 : undefined)
+        }
+    }, [selectors, selectorMatches, checkUniqueness])
```
Comment (mariusandra) [inline frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.tsx:117]: This already looks like enough internal state to put in a logic (with a simple unique key)

**Review by mariusandra — COMMENTED** [review]
Nicely done! 🙌 

Some code nits inline, though I think there's one thing missing from a UX perspective to make it spark joy:

1. When I click "edit", I expect to be able to edit the existing selector, but I start with a blank slate:


2. It feels natural to be able to select more than one class, but I can just choose one:


3. This selector line could use a monospaced font, and look a bit more code-like.
image

File: frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.scss (hunk 1 above)
Comment (pauldambra) [reply frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.scss:1]: Took me ages to track down usage (or at least way longer than needed) -> renamed for clarity

**Comment by pauldambra** [comment]
Addressed most of that... (I think)

Screenshot 2023-02-15 at 23 36 37

Screenshot 2023-02-15 at 23 38 05

> When I click "edit", I expect to be able to edit the existing selector, but I start with a blank slate:

Yep, that's also pretty tricky 🤣 At the minute this works because it's pretty simple and isn't really HTML aware. Ideally, we can carry on without that. And then if that's all users ask for you can have a free "I told you so"

File: frontend/src/toolbar/elements/elementsLogic.ts [hunk 3]
```diff
@@ -32,7 +33,10 @@ export const elementsLogic = kea<elementsLogicType>({
         enableInspect: true,
         disableInspect: true,
 
-        selectElement: (element: HTMLElement | null) => ({ element }),
+        selectElement: (element: HTMLElement | null, elementType?: 'heatmap' | 'inspect' | 'action') => ({
```
Comment (mariusandra) [inline frontend/src/toolbar/elements/elementsLogic.ts:36]: Nit, it's not really element "type" here, but more like the click's "source". The element's type is still a HTML Link/Button/etc Element.

File: frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                    />
+                )
+            })}
+        </>
+    )
+}
+
+let uniqueNode = 0
+
+export function HtmlElementsDisplay({
```
Comment (mariusandra) [inline frontend/src/lib/components/HtmlElementsDisplay/HtmlElementsDisplay.tsx:64]: We usually don't PascalCase abbreviations, so it should be `HTMLElementsDisplay`

**Review by mariusandra — CHANGES_REQUESTED** [review]
> > When I click "edit", I expect to be able to edit the existing selector, but I start with a blank slate:
> 
> Yep, that's also pretty tricky 🤣 At the minute this works because it's pretty simple and isn't really HTML aware. Ideally, we can carry on without that. And then if that's all users ask for you can have a free "I told you so"

Hmm... 

I applaud you for taking on this task, but if my request of "can it preselect the existing value" is met with "that's very tricky", then I'm going to conclude that the code is poorly written and should be redone. Especially when I find new code with comments like "Yuck warning".

Going through the code, it's also very hard to review, as many tiny things changed to pass state over the logics and components. There's new useState component state, there's new logic state. Things are sometimes overriding other things (`newActionSelectorOverride`, `overriddenSelectors`). E.g. I'm not sure why anything needs to be "overridden" in the first place, as this "selector string" is just a string that we send in a form 🤷‍♂️. In fact it would be fine if the only place you can edit the selector is the form when creating an action. The only place where you can't edit the selector is the form where you're supposed to edit the selector. That doesn't make sense.

Thus:
1. Can you remove the "edit" button next to the selector in the infowindow. You needed to write a lot of tech debt to make that work, and I'm not sure it's the right place for this. Make that pencil open the "add new action" view if you want to keep it there, but I think we're fine to remove it.
2. Can you add the element selector to the add/edit action form.
3. It's an edit form. There's no way we can't preselect selected values. Can you make it work? Throw out all the old code and rewrite it from first principles to support this if you need to.

If you'd rewrite this from the beginning, what would you do differently? Can we get this to a state where we can merge it in without a lot of new tech dept? Can this "HTMLElementsDisplay" be a totally pure component that only accepts "value" (selector string), "onChange" (str => void), "elementsChain" and "DOM element in which to search for matches".

**Comment by pauldambra** [comment]
> "that's very tricky", then I'm going to conclude that the code is poorly written and should be redone

I think it's a genuinely hard problem - things can be tricky in and of themselves 🤷. Writing this was specifically to let the human brain deal with the complexity without us parsing HTML/CSS.

**Comment by pauldambra** [comment]
now with edit flow

I think we'll get away with not being exposed to the full complexity of CSS because we're dealing with a pretty constrained case (a single element and up to ten of its ancestors) so we don't need to handle all of the combinators and pseudo selectors

File: frontend/src/lib/components/HTMLElementsDisplay/htmlElementsDisplayLogic.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+export const htmlElementsDisplayLogic = kea<htmlElementsDisplayLogicType>([
+    path(['lib', 'components', 'HtmlElementsDisplay', 'htmlElementsDisplayLogic']),
+    props({ checkUniqueness: true } as HtmlElementDisplayLogicProps),
+    key((props) => props.key),
+    actions({
+        setParsedSelectors: (selectors: Record<number, ParsedCSSSelector>) => ({ selectors }),
+        // chooseSelector: (chosenSelector: ChosenSelector) => ({
+        //     chosenSelector,
+        // }),
```
Comment (mariusandra) [inline frontend/src/lib/components/HTMLElementsDisplay/htmlElementsDisplayLogic.ts:34]: `//`

File: frontend/src/lib/components/HTMLElementsDisplay/preselectWithCSS.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                parts[processing] = current
+            }
+            current = ''
+        }
+    }
+
+    // processing things with a grammar like html and css with regex is a route to madness
+    // pulling in a library like parsel is taking on a new dependency for a hopefully limited use case
+    // we don't need to support all the css selectors,
+    // so, we'll just do it manually (until we need the new dependency)
```
Comment (mariusandra) [inline frontend/src/lib/components/HTMLElementsDisplay/preselectWithCSS.ts:60]: ❤️

**Review by mariusandra — COMMENTED** [review]
This is looking great! Some points I could find, not all needed for this PR.

1. I found something vaguely interesting:



The selector that appears is different than what it shows initially. I'm not sure if that's related to this PR or not, but might be a quick fix that can be done here.

2. For another quick ~scope creep~ win, can you make the "selector" the first element in the form? It feels like the "edit the selector" and "change element" buttons deserve to be somewhat closer? Context: When I found point nr 1, I wanted to select a new element, but couldn't find any button next to the new "edit selector" button that lets me do that... So I even though you had removed it, but then found the "change element" button above. 

3. Visual regression test failing with the new view you added :).

4. Somehow half of the elements view has a dark grey background, and the other half has a black background. Might look nicer in just one color...
image

5. The :nth-element() bit is missing, but I'm not sure how to add that. We can do it another time/PR, though for feature completeness, I think the easiest way is to just add a "modifier" for each tag, either at the beginning or end of each tag line, which you could just click and edit in place (`window.prompt` or a lemon inline popup/confirm). I don't think that free text at the end of each tag is anything complicated after everything you've been through, but feel free to hit the "defer" button on this 😅 

image

**Comment by pauldambra** [comment]
> The selector that appears is different than what it shows initially. I'm not sure if that's related to this PR or not, but might be a quick fix that can be done here.

Yep, two years old 😓 -> but also, actually, a quick fix 🏅 

> can you make the "selector" the first element in the form? 

✅ 

> Somehow half of the elements view has a dark grey background, and the other half has a black background. 

Only has black background when the leaf element is highlighted now (so the events table doesn't change)

-----

I'll hit defer on the nth-child stuff, and PRs are welcome 😅 

If you have `div div:nth-child(3) div` and our element chain is `<div> <div> <div> <div> <div> <button/> <closingtags/>` we need to be aware (maybe) of the rest of the DOM to know which div could be the nth-child matcher. I might be missing an easy way but we're at the magic 55 commits level where it would be good to get it in and then iterate

**Review by mariusandra — APPROVED** [review]
I had a play around... and it looks and works really nicely! Well done! 👍 ✅

**Pull request merged**